File: src/main/java/no/mesan/fag/patterns/timesheet/external/TimeIteratorService.java
package no.mesan.fag.patterns.timesheet.external;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import no.mesan.fag.patterns.timesheet.data.TimesheetEntry;

/** Adapter TimeDataService til en TimeServiceAdapter. */
public class TimeIteratorService implements TimeServiceAdapter {
    private final TimeDataService original;

    public TimeIteratorService(final TimeDataService original) {
        this.original = original;
    }

    private abstract static class EntryIterator implements Iterator<TimesheetEntry> {
        /** Implementer denne for å hente data. */
        protected abstract List<TimesheetEntry> getNextFrom(final int from);

        // Holder de vi har hentet så langt
        private final List<TimesheetEntry> buffer= new ArrayList<>();

        // Antall vi har plukket ut
        int fetched= 0;

        // Har vi hentet alle?
        boolean lastIsFetched= false;

        private void refillIfNeeded() {
            if (!buffer.isEmpty() || lastIsFetched) return;
            final List<TimesheetEntry> newBatch = getNextFrom(fetched);
            fetched += newBatch.size();
            buffer.addAll(newBatch);
            lastIsFetched = newBatch.isEmpty();
        }

        /// HINT Du må lage hasNext() og next() for at dette skal virke...
        ///      datastrukturene og metoden over er ment å kunne hjelpe deg med dette

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
    }

    @Override
    public Iterable<TimesheetEntry> forEmployee(final String userID) {
        return null; /// HINT returner en lambda basert på klassen over, omtrent slik:
                     ///    return () -> new EntryIterator() { ... }
    }

    @Override
    public Iterable<TimesheetEntry> forYear(final int year) {
        return null;  /// HINT ditto
    }
}

File: src/main/scala/no/mesan/fag/patterns/scala/timesheet/external/TimeIteratorService.scala
package no.mesan.fag.patterns.scala.timesheet.external

import no.mesan.fag.patterns.scala.timesheet.data.TimesheetEntry

/** Adapter til den eksterne tjenesten. */
class TimeIteratorService(source: TimeDataService) extends TimeServiceAdapter {

  private class EntryIterator(fillFunction: Int=>List[TimesheetEntry]) extends Iterator[TimesheetEntry] {
    var buffer: List[TimesheetEntry] = Nil
    var fetched = 0
    var lastIsFetched = false

    override def hasNext: Boolean = {
      refillIfNeeded
      !buffer.isEmpty
    }

    private def refillIfNeeded {
      if (!buffer.isEmpty || lastIsFetched) return
      val newBatch: List[TimesheetEntry] = fillFunction(fetched)
      fetched += newBatch.size
      buffer ++= newBatch
      lastIsFetched = newBatch.isEmpty
    }

    override def next(): TimesheetEntry = {
      refillIfNeeded
      if (buffer.isEmpty) throw new NoSuchElementException
      val entry= buffer.head
      buffer= buffer.tail
      entry
    }
  }

  override def forEmployee(userID: String): Iterable[TimesheetEntry] =
    new EntryIterator(fetched => source.forEmployee(userID, fetched)).toIterable

  override def forYear(year: Int): Iterable[TimesheetEntry]=
    new EntryIterator(fetched => source.forYear(year, fetched)).toIterable
}
